File: leetcode/89/格雷编码.js
/*
 * @lc app=leetcode.cn id=89 lang=javascript
 *
 * [89] 格雷编码
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number[]}
 */
var grayCode = function (n) {
  if (n == 0) {
    return [0];
  }
  if (n == 1) {
    return [0, 1];
  }

  let res = [0, 1];
  for (let i = 1; i < n; i++) {
    let temp = [...res];
    let reverse = res.reverse();
    reverse = reverse.map(el => el + (1 << i));
    res = [...temp, ...reverse];
  }

  return res;
};
// @lc code=end

File: leetcode/89/格雷编码.java
/*
 * @lc app=leetcode.cn id=89 lang=java
 *
 * [89] 格雷编码
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<Integer> grayCode(int n) {
        ArrayList<Integer> res = new ArrayList<Integer>();
        for (int i = 0; i < 1 << n; i++) {
            res.add(i ^ (i >> 1));
        }

        return res;
    }
}
// @lc code=end
